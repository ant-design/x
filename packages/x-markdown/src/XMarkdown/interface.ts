import type { MarkedExtension, MarkedOptions, Tokens } from 'marked';
import { CSSProperties } from 'react';
import { Renderer } from './core';

export interface MarkdownXOptions extends MarkedOptions {
  XRenderer: Renderer;
}

export type Token = Tokens.Generic;

export interface SteamingOption {
  /**
   * @description 是否还有流式数据
   * @default false
   */
  hasNextChunk: boolean;
}

export interface Config {
  break?: boolean;
  gfm?: boolean;
}

export type GenericRendererFunction = (...args: unknown[]) => React.ReactNode | false;

export type RendererObject = {
  [key: string]: GenericRendererFunction;
};

export interface plugin extends Omit<MarkedExtension, 'renderer' | 'walkTokens'> {
  renderer?: any;
  walkTokens?: (token: Token) => void;
}

export interface XMarkdownProps {
  content?: string;
  children?: string;
  config?: Config;
  allowHtml?: boolean;
  components?: Record<string, any>;
  streaming?: SteamingOption;
  plugins?: plugin[];
  className?: string;
  style?: CSSProperties;
}
