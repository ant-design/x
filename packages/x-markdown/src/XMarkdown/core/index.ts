import Lexer from './Lexer';
import { createMarkdownOptions } from './Options';
import Parser from './Parser';
import Renderer from './Renderer';

export { createMarkdownOptions, Lexer, Parser, Renderer };
