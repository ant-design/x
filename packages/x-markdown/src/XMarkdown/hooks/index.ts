import useStreaming from './useStreaming';
import useAnimation from './useAnimation';

export { useStreaming, useAnimation };
