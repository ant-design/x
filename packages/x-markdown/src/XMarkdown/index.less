// XMarkdown default styles
// Note: No theme colors included, only basic layout and typography styles

.x-markdown {
  // Base styles
  line-height: 1.5714285714285714;
  word-wrap: break-word;
  overflow-wrap: break-word;

  // Long content wrapping
  p,
  div,
  span,
  li,
  td,
  th {
    word-break: break-word;
    overflow-wrap: break-word;
  }

  // Code block wrapping
  pre,
  code {
    word-break: break-word;
    overflow-wrap: break-word;
    white-space: pre-wrap;
  }

  // Table overflow handling
  table {
    word-break: break-word;
    overflow-wrap: break-word;
    table-layout: fixed;
    width: 100%;
  }

  // Paragraph styles optimization
  p {
    margin: 0 0 1em 0;

    // First paragraph has no top margin
    &:first-child {
      margin-top: 0;
    }

    // Last paragraph has no bottom margin
    &:last-child {
      margin-bottom: 0;
    }
  }

  // List styles optimization
  ul,
  ol {
    margin: 0 0 1em 0;
    padding-left: 2em;

    &:first-child {
      margin-top: 0;
    }

    &:last-child {
      margin-bottom: 0;
    }
  }

  li {
    margin: 0.25em 0;

    &:first-child {
      margin-top: 0;
    }

    &:last-child {
      margin-bottom: 0;
    }
  }

  // Code block styles
  pre {
    margin: 1em 0;
    overflow-x: auto;
    &:first-child {
      margin-top: 0;
    }

    &:last-child {
      margin-bottom: 0;
    }
  }

  // Inline code styles
  code {
    padding: 0.2em 0.4em;
    margin: 0;
    font-size: 0.85em;
    border-radius: 3px;
  }

  pre code {
    padding: 0;
    margin: 0;
    font-size: inherit;
    border-radius: 0;
    line-height: 2;
  }

  // Image styles
  img {
    max-width: 100%;
    height: auto;
    margin: 0.5em 0;
  }

  // Horizontal rule
  hr {
    margin: 1.5em 0;
    border: none;
    border-top: 1px solid #ddd;
  }

  // Table styles
  table {
    margin: 1em 0;
    border-collapse: collapse;

    &:first-child {
      margin-top: 0;
    }

    &:last-child {
      margin-bottom: 0;
    }
  }

  th,
  td {
    padding: 0.5em;
    border: 1px solid #ddd;
    text-align: left;
  }

  th {
    font-weight: bold;
  }
}
