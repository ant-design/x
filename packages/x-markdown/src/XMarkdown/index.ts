export { default, default as XMarkdown } from './index.tsx';

export { Lexer, Parser, Renderer } from './core';

export type { MarkdownProps } from './interface.ts';
