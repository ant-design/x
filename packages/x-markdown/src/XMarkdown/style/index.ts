import { mergeToken } from '@ant-design/cssinjs-utils';
import { genStyleHooks } from '../../theme/genStyleUtils';
import type { GenerateStyle } from 'antd/es/theme/internal';
import type { GetDefaultToken, FullToken } from '../../theme/useToken';

export interface ComponentToken {}

export interface XMarkdownToken extends FullToken<'Mermaid'> {}

const genXMarkdownStyle: GenerateStyle<XMarkdownToken> = (token) => {
  const {
    componentCls,
    paddingXS,
    paddingSM,
    marginSM,
    marginLG,
    marginMD,
    margin,
    fontSizeHeading3,
    fontSizeHeading4,
    fontSizeHeading5,
    paddingXXS,
    lineHeightLG,
    lineHeight,
  } = token;

  return {
    [componentCls]: {
      fontSize: fontSizeHeading5,
      lineHeight,

      '& h1, & h2, & h3, & h4, & p, & pre': {
        marginBottom: margin,
      },

      '& p+h1': {
        marginTop: marginLG,
      },

      '& p+h2': {
        marginTop: marginMD,
      },

      '& p+h3, & p+h4': {
        marginTop: margin,
      },

      '& h1': {
        fontSize: fontSizeHeading3,
        fontWeight: 500,
        lineHeight: lineHeightLG,
      },

      '& h2': {
        fontSize: fontSizeHeading4,
        fontWeight: 500,
        lineHeight: lineHeightLG,
      },

      '& h3': {
        fontSize: 18,
        fontWeight: 500,
        lineHeight: lineHeightLG,
      },

      '& h4': {
        fontSize: fontSizeHeading5,
        fontWeight: 500,
      },

      '& p': {
        fontSize: fontSizeHeading5,
      },

      '& hr': {
        marginBottom: marginLG,
        marginTop: marginLG,
      },

      '& code': {
        backgroundColor: 'transparent',
        display: 'inline-flex',
      },

      '& ul, & ol': {
        marginBottom: marginSM,
        marginTop: paddingXS,
        paddingLeft: paddingSM,
        textAlign: 'left',
      },

      '& ul': {
        listStyle: 'disc',
        '& ul': {
          listStyle: 'circle',
          '& ul': {
            listStyle: 'square',
          },
        },
      },

      '& ol': {
        listStyle: 'decimal',
      },

      '& li': {
        marginBottom: paddingXXS,
        marginLeft: paddingXXS,
        wordBreak: 'break-word',
        wordWrap: 'break-word',
        overflowWrap: 'break-word',
      },
    },
  };
};

export const prepareComponentToken: GetDefaultToken<'XMarkdown'> = () => ({});

export default genStyleHooks<'XMarkdown'>(
  'XMarkdown',
  (token) => {
    const xmarkdownToken = mergeToken<XMarkdownToken>(token, {});
    return [genXMarkdownStyle(xmarkdownToken)];
  },
  prepareComponentToken,
);
