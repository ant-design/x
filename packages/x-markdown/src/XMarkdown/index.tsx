import { Marked, walkTokens } from 'marked';
import React from 'react';
import { Parser, createMarkdownConfig } from './core';
import useStreaming from './hooks/useStreaming';
import { XMarkdownProps } from './interface';

const XMarkdown: React.FC<XMarkdownProps> = (props) => {
  const {
    content,
    children,
    options,
    streaming,
    plugins,
    components,
    walkTokens: csWalkToken,
    className,
    style,
  } = props;

  // ============================ Render ============================
  const displayContent = useStreaming(content || children || '', streaming, components);
  if (!displayContent) return null;

  const config = createMarkdownConfig(options, plugins, components, streaming);

  const markedInstance = new Marked({
    breaks: config.break,
    gfm: config.gfm,
    extensions: config.plugins,
  });

  if (markedInstance.defaults.extensions) {
    config.extensions = markedInstance.defaults.extensions;
  }

  const parser = new Parser(config);

  const tokens = markedInstance.lexer(displayContent);

  console.log('tokens', tokens);

  if (csWalkToken) {
    walkTokens(tokens, csWalkToken);
  }
  return (
    <div className={className} style={style}>
      {parser.parse(tokens)}
    </div>
  );
};

if (process.env.NODE_ENV !== 'production') {
  XMarkdown.displayName = 'XMarkdown';
}

export default XMarkdown;
