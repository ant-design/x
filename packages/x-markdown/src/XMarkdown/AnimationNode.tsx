import React, { useEffect, useRef, useState } from 'react';
import { HTMLTag } from './hooks/useAnimation';
import { AnimationConfig } from './interface';

export interface AnimationNodeProps extends React.HTMLAttributes<HTMLElement> {
  children?: React.ReactNode | React.ReactNode[];
  nodeTag: HTMLTag;
  animationConfig?: AnimationConfig;
  [key: string]: unknown;
}

export interface AnimationTextProps {
  text: string;
  animationConfig?: AnimationConfig;
}

const AnimationText = React.memo<AnimationTextProps>((props) => {
  const { text, animationConfig } = props;
  const { fadeDuration = 200, easing = 'ease-in-out' } = animationConfig || {};
  const [chunks, setChunks] = useState<string[]>([]);
  const prevTextRef = useRef('');

  useEffect(() => {
    if (text === prevTextRef.current) return;

    if (!(prevTextRef.current && text.startsWith(prevTextRef.current))) {
      setChunks([text]);
      prevTextRef.current = text;
      return;
    }

    const newText = text.slice(prevTextRef.current.length);
    if (!newText) return;

    setChunks((prev) => [...prev, newText]);
    prevTextRef.current = text;
  }, [text]);

  return (
    <>
      {chunks.map((text, index) => {
        return (
          <span
            style={{
              animation: `fadeIn ${fadeDuration}ms ${easing} forwards`,
            }}
            key={`${index}-${text}`}
          >
            {text}
          </span>
        );
      })}
      <style>{`
        @keyframes fadeIn {
          from { opacity: 0; }
          to { opacity: 1; }
        }
      `}</style>
    </>
  );
});

const AnimationNode: React.FC<AnimationNodeProps> = (props) => {
  const { nodeTag, children, animationConfig, _domNode, _streamStatus, ...restProps } = props;

  const renderChildren = (): React.ReactNode | React.ReactNode[] => {
    if (!children) return null;

    if (Array.isArray(children)) {
      return children.map((child, index) =>
        typeof child === 'string' ? (
          <AnimationText key={index} animationConfig={animationConfig} text={child} />
        ) : (
          child
        ),
      );
    }
    return typeof children === 'string' ? (
      <AnimationText text={children} animationConfig={animationConfig} />
    ) : (
      children
    );
  };

  return React.createElement(nodeTag, restProps, renderChildren());
};

export default AnimationNode;
export { AnimationText };
