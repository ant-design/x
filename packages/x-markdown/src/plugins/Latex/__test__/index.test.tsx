import { render } from '@testing-library/react';
import React from 'react';
import XMarkdown from '../../../XMarkdown';

// Mock CSS import to avoid Jest issues
jest.mock('katex/dist/katex.min.css', () => ({}));

// Mock katex to return HTML instead of MathML
jest.mock('katex', () => ({
  renderToString: (text: string, options: any) => {
    const displayClass = options?.displayMode ? 'katex-display' : 'katex';
    return `<span class="${displayClass}"><span class="katex-mathml">${text}</span><span class="katex-html" aria-hidden="true">${text}</span></span>`;
  },
}));

// Import the actual plugin after mocking
import latexPlugin from '../index';

describe('LaTeX Plugin', () => {
  it('should render inline LaTeX with $..$ syntax', () => {
    const { container } = render(
      <XMarkdown config={{ extensions: latexPlugin() }}>{'$E=mc^2$'}</XMarkdown>,
    );

    expect(container.innerHTML).toContain('katex');
    expect(container.innerHTML).toContain('E=mc^2');
  });

  it('should render block LaTeX with $$..$$ syntax', () => {
    const { container } = render(
      <XMarkdown config={{ extensions: latexPlugin() }}>{'$$\\frac{a}{b}$$'}</XMarkdown>,
    );

    expect(container.innerHTML).toContain('katex');
    expect(container.innerHTML).toContain('\\frac{a}{b}');
  });

  it('should render block LaTeX with \\[..\\] syntax', () => {
    const { container } = render(
      <XMarkdown config={{ extensions: latexPlugin() }}>{'\\[\\frac{a}{b}\\]'}</XMarkdown>,
    );

    expect(container.innerHTML).toContain('katex');
    expect(container.innerHTML).toContain('\\frac{a}{b}');
  });

  it('should handle LaTeX with surrounding text', () => {
    const { container } = render(
      <XMarkdown config={{ extensions: latexPlugin() }}>
        {'This is an equation: $E=mc^2$ in text'}
      </XMarkdown>,
    );

    expect(container.innerHTML).toContain('katex');
    expect(container.innerHTML).toContain('E=mc^2');
  });

  it('should handle multiple LaTeX formulas', () => {
    const { container } = render(
      <XMarkdown config={{ extensions: latexPlugin() }}>{'$a+b$ and $$\\frac{c}{d}$$'}</XMarkdown>,
    );

    expect(container.innerHTML).toContain('katex');
    expect(container.innerHTML).toContain('a+b');
    expect(container.innerHTML).toContain('\\frac{c}{d}');
  });

  it('should handle align* syntax replacement', () => {
    const { container } = render(
      <XMarkdown config={{ extensions: latexPlugin() }}>
        {'$$ \\begin{align*} x &= y \\ y &= z \\end{align*} $$'}
      </XMarkdown>,
    );

    expect(container.innerHTML).toContain('katex');
    expect(container.innerHTML).toContain('aligned');
  });

  it('should handle empty content', () => {
    const { container } = render(
      <XMarkdown config={{ extensions: latexPlugin() }}>{''}</XMarkdown>,
    );

    expect(container).toBeTruthy();
  });

  it('should handle content without LaTeX', () => {
    const { container } = render(
      <XMarkdown config={{ extensions: latexPlugin() }}>{'Just plain text'}</XMarkdown>,
    );

    expect(container.innerHTML).not.toContain('katex');
  });

  it('should handle complex LaTeX expressions', () => {
    const { container } = render(
      <XMarkdown config={{ extensions: latexPlugin() }}>
        {'$\\sum_{i=1}^{n} x_i = \\prod_{j=1}^{m} y_j$'}
      </XMarkdown>,
    );

    expect(container.innerHTML).toContain('katex');
    expect(container.innerHTML).toContain('\\sum_{i=1}^{n} x_i');
  });

  it('should handle mixed LaTeX syntaxes', () => {
    const { container } = render(
      <XMarkdown config={{ extensions: latexPlugin() }}>
        {'Inline: $x^2$ and block: $$\\int_0^1 f(x)dx$$'}
      </XMarkdown>,
    );

    expect(container.innerHTML).toContain('katex');
    expect(container.innerHTML).toContain('x^2');
    expect(container.innerHTML).toContain('\\int_0^1 f(x)dx');
  });
});
