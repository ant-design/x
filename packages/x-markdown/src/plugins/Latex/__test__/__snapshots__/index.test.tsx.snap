// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaTeX Plugin testcase: should render block LaTeX with $$..$$ syntax 1`] = `
<div>
  <div>
    <p>
      This is an block LaTeX: 
    </p>
    <p>
      <span
        class="katex"
      />
    </p>
  </div>
</div>
`;

exports[`LaTeX Plugin testcase: should render block LaTeX with \\[..\\] syntax 1`] = `
<div>
  <div>
    <p>
      This is an block LaTeX: 
    </p>
    <pre>
      <code>
          \\[
      </code>
    </pre>
    <p>
      \\Delta t' = \\frac{\\Delta t}{\\sqrt{1 - \\frac{v^2}{c^2}}}
      ]
    </p>
  </div>
</div>
`;

exports[`LaTeX Plugin testcase: should render inline LaTeX with $..$ syntax 1`] = `
<div>
  <div>
    <h3>
      Latex
    </h3>
    <p>
      inline standard: 
      <span
        class="katex"
      />
    </p>
    <p>
      block standardï¼š
    </p>
    <span
      class="katex"
    />
    <p>
      inline: 
      <span
        class="katex"
      />
    </p>
    <p>
      block: 
    </p>
    <span
      class="katex"
    />
  </div>
</div>
`;

exports[`LaTeX Plugin testcase: should render inline LaTeX with \\(..\\) syntax 1`] = `
<div>
  <div>
    <p>
      This is an inline LaTeX: 
      <span
        class="katex"
      />
    </p>
  </div>
</div>
`;
